Variable/Value operations before evaluation

derekpierre opened the pull request:
**Type of PR:**
- [ ] Bugfix
- [x] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:** 
- [ ] 1
- [x] 2
- [ ] 3

**What this does:**
- Allow capability to perform transformations on variables/values for `SequentialCondition` and `ReturnValueTest` before evaluation.
- This capability is provided via the `VariableOperation` object associated with `ConditionVariable` (`SequentialCondition`) and `ReturnValueTest` 

**Issues fixed/closed:**
> - Fixes #...

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
> What should reviewers focus on? 
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?

derekpierre committed:
Add simple unit test to check whether our existing signing conditions can limit the amount to transfer of an erc20 token based on points/values obtained from a JSON REST server.

Co-authored-by: Manuel Montenegro <manuel@nucypher.com>

derekpierre committed:
Initial musing about how we can potentially incorporate operations on data used for conditions.

derekpierre committed:
Add "index" variable operation.

derekpierre committed:
Add round, ceil, floor operations.

derekpierre committed:
Re-affirm that we are good with JSON queries returning multiple values and not forcing just one.
Remove commented out code and update tests to allow multiple values being returned.

derekpierre committed:
Code cleanup: validate schema in constructor, catch and reraise exceptions during use of variable operations, raise type error instead of value error for operation values.

derekpierre committed:
Add initial unit tests for variable operation(s).

derekpierre committed:
Add operations for abs, and casting.
Reorder operator functions to be alphabetical.
Update unit tests.

derekpierre committed:
Add tests for usage of VariableOperations with ReturnValueTest and SequentialCondition (via ConditionVariable).

derekpierre committed:
Use AnyField instead of Raw to account for the case when `taco-web` provides large numbers (bigints) as strings.

derekpierre committed:
Validate ConditionVariable when created since it derives from _Serializable. Without it, constructor params are not properly validated and instead only as part of a condition schema.

derekpierre committed:
Remove use of VariableOperations in favour of List[VariableOperation]; this will clean up the serialized json of the operations field used in ReturnValueTest and ConditionVariable.
Define static function to easily evaluate list of variable operations.

derekpierre committed:
Define proper types for VariableOperation and its usage.
Fix incorrect field for ReturnValueTest - "key" should be "index".

derekpierre committed:
Remove use of VariableOperations from tests now that it was removed; instead use List[VariableOperation].

derekpierre committed:
Add tests for serialization/deserialization of operations to lingo tests.

derekpierre committed:
Update typing import based on python version now that we only support python 3.10 and newer.

derekpierre committed:
Don't explicitly check types for operations now that we expect TypeErrors - the underlying operation will throw their own TypeError.

derekpierre committed:
Improve validation of operation and value combination.
Update tests.

derekpierre committed:
Add support for context variable resolution for VariableOperation - the "value" property could be provided as a context variable.

derekpierre committed:
Add higher-level support for converting to/from wei.
Update unit test.

derekpierre committed:
Pare back operators to exclude "median" and "mode"; seems excessive and less useful. If needed, we can always add them back at a later time.
Reorder alphabetically.

derekpierre committed:
Add feature newsfragment for #3658.

Muhammad-Altabba reviewed (commented):
I am not much into python. But this sounds great!

Muhammad-Altabba commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -1,11 +1,15 @@
 import ast
```
I think what Copilot suppressed due to low confidence is valid

> <summary>Comments suppressed due to low confidence (1)</summary>
> 
> **nucypher/policy/conditions/lingo.py:1**
> * The import of `RequiredContextVariable` is unused in this file. Consider removing it to keep imports clean.
> ```
> import ast
> ```

derekpierre replied:
Perhaps I'm misunderstanding, but it's unclear what the issue is. `lingo.py` does not have an import for `RequiredContextVariable`. If you have an idea, I'm all ears.

Muhammad-Altabba replied:
Sorry, I was confused with something else. And there is no such issue.

theref commented on nucypher/policy/conditions/json/utils.py:
```diff
@@ -53,11 +50,8 @@ def query_json_data(data: Any, query: Optional[str], **context) -> Any:
-        )
-        logger.info(message)
-        raise JsonRequestException(message)
+        result = [match.value for match in matches]
```
how come we're changing this?

derekpierre replied:
Previously, we allowed only 1 result to be returned from a Json endpoint, because the condition could only check the value returned.  Now, because of operations, we can actually check a value based on multiple outputs from the JSON because we can perform operations on it (eg. `sum`). 

The cleanest example is this test -  https://github.com/derekpierre/nucypher/blob/variable-operations/tests/unit/conditions/test_signing_condition.py#L862 - which mimics the Engages API Rest endpoint. This is case where the user has points attributed to the different guilds they belong to, and the amount of tokens they can claim in the TGE is a function of the total points they have (in this test it is total points == allowed tokens in wei in the UserOp)

theref replied:
love it, makes complete sense

theref commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -291,18 +296,171 @@ def __init__(self, operand: Condition):
+    "*=": pyoperator.mul,
+    "/=": pyoperator.truediv,
+    "%=": pyoperator.mod,
+    "^=": pyoperator.pow,
```
are we doing anything about overly "expensive" conditions?

derekpierre replied:
✔️ - good question. It is now limited to a max of 5 operations.

theref replied:
it's actually something like this that worries me `[{"operation": "^=", "value": 999999}` - might be worth testing to see if it errors gracefully or crashes

KPrasch replied:
☝🏻 Valid concern - even if memory holds, Python will keep crunching a ** b until timeout. CPU time grows super-linearly with the result size.  some options to consider:

1) `pow(a, b, m)` (specifically the three-arg usage) is the safe form (O(log b) time, constant memory), but not usable for exact equality comparison 
2) Add a pre-check to estimate digits and bail if it’s too large
3) Simplest fix: disable exponent transforms for now until we define safe limits or a modular form

theref replied:
another option is a timeout:
```python
@contextmanager
def timeout(seconds=1):
    """Context manager that raises OperationTimeout if code takes too long."""
    def timeout_handler(signum, frame):
        raise OperationTimeout(f"Operation exceeded {seconds} second timeout")
    
    # Set the signal handler
    old_handler = signal.signal(signal.SIGALRM, timeout_handler)
    signal.alarm(seconds)
    try:
        yield
    finally:
        signal.alarm(0)
        signal.signal(signal.SIGALRM, old_handler)
```
expensive operations could then be wrapped:
```python
class VariableOperation(_Serializable):
    OPERATION_TIMEOUT_SECONDS = 1  # 1 second max per operation
    
    def calc(self, variable_value: Any):
        """Calculate the result of the operation on the variable value."""
        operation_function = _OPERATOR_FUNCTIONS[self.operation]
        
        try:
            with timeout(self.OPERATION_TIMEOUT_SECONDS):
                return operation_function(variable_value, self.value)
        except OperationTimeout as e:
            raise ConditionEvaluationFailed(
                f"Operation '{self.operation}' with value '{self.value}' exceeded timeout. "
                f"Result would be too expensive to compute."
            ) from e
```
the timeout is only be applied to a single operation function. And any single maths operation that's taking >1-2 secs is sus

Muhammad-Altabba replied:
If putting a timeout, I think it is better to set it on the root condition evaluation.

derekpierre replied:
Great convo.

✔️ - Removed for now since no one has really asked for this functionality, and unclear of its usefulness at the moment given the expressed concern.

theref commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -291,18 +296,171 @@ def __init__(self, operand: Condition):
+    "str": lambda a, _: str(a),
+}
+
+_OPERATORS_WITH_NO_VALUE = {
```
what do you mean "no value"?

derekpierre replied:
It doesn't take an additional parameter i.e. no `b` value for the lambda function.

✔️ Renamed to `UNARY_OPERATOR_FUNCTIONS`.

theref commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -709,7 +879,29 @@ def eval(self, data) -> bool:
+                    f"Error performing operations on returned data: {e}"
+                )
+
+        processed_data = self._process_data(data)
```
i'm unsure about processing after operations. Don't we need to convert types and then do operations?

derekpierre replied:
The `_process_data` is the final step before evaluation/comparison. This step is for internal use and is specifically to facilitate `ast.literal_eval` when comparing byte values, and is invisible to the caller (we don't document this). That's why all operations are performed on the raw values returned from the condition (which the caller is aware of) before this internal processing takes place.

theref commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -671,31 +856,16 @@ def _sanitize_value(cls, value):
         return HexBytes(data).hex() if isinstance(data, bytes) else data
 
-    def _process_data(self, data: Any, index: Optional[int]) -> Any:
+    def _process_data(self, data: Any) -> Any:
```
Can you explain why index is removed?

derekpierre replied:
We need to pull out the value at index earlier before the processing since it may need to undergo operations first. This is done on L823.

This `_process_data` is for modifying the value (bytes to hex, perhaps others in the future) before the comparison is performed.

Updated the function description.

derekpierre committed:
Add restriction on maximum number of operations that can be performed on a variable/result.
Update tests accordingly.

derekpierre committed:
Rename constant for better clarity.

derekpierre committed:
Clean up method description since index is no longer provided to function.

derekpierre reviewed (commented):
Great questions.

cygnusv commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -291,18 +296,180 @@ def __init__(self, operand: Condition):
+        raise TypeError(f"Invalid value for ethToWei conversion: {a}") from e
+
+
+def _wei_to_eth(a, _):
```
```suggestion
def _wei_to_eth(a, _) -> decimal.Decimal:
```

derekpierre replied:
Fixed via https://github.com/derekpierre/nucypher/pull/9

cygnusv commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -291,18 +296,180 @@ def __init__(self, operand: Condition):
 }
 
 
+def _eth_to_wei(a, _):
```
```suggestion
def _eth_to_wei(a, _) -> int:
```

derekpierre replied:
Fixed via https://github.com/derekpierre/nucypher/pull/9

cygnusv committed:
Add failing tests due to float precision issues

cygnusv committed:
Add some text examples for overloaded operators

cygnusv commented:
We have to be very careful with what operations we allow here. There's many potential traps (e.g. floating point precision issues, overloaded operators, etc) that we need to weigh very carefully. Please see my branch https://github.com/cygnusv/nucypher/tree/variable-operations, and in particular the last 2 commits with some examples.

cygnusv committed:
Define return type from currency util functions

cygnusv committed:
Relocate currency helpers to utils module

cygnusv committed:
Add @pre_load processing to convert float to decimal

cygnusv committed:
Some clarifications to test cases for VariableOperation

cygnusv committed:
Preprocessing steps don't kick in when constructor is used directly

cygnusv commented:
Created a PR to Derek's branch with some ideas, but I think this requires discussion with Derek:
https://github.com/derekpierre/nucypher/pull/9

derekpierre committed:
Remove ^= (pow) operator due to expensive nature of calculation and possible ddos vector.

derekpierre committed:
Convert from float to decimal in constructor so that conversion is always performed, and not just from json.
Decimal is not JSON serializable and is really an internal structure used for calcs, convert from Decimal back to float at the end of the operations.
Force use of calc_from_list() to ensure precision is maintained across all calcs, and make _calc() private.
Update tests and add test to ensure that ReturnValueTest usage of VariableOperation properly performs comparisons.

manumonti commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -291,18 +296,180 @@ def __init__(self, operand: Condition):
+                    raise ValidationError(
+                        field_name="value",
+                        message=f'No value should be provided for operation "{operation}"',
+                    )
```
I have a suggestion here. Instead of having a list of operators-without-2nd-value, which sounds to me like code-redundant, we could do something like this. (I have not tested it, though).

```suggestion
from inspect import signature  # at the begginning of the file 

class VariableOperation(_Serializable):
    class Schema(CamelCaseSchema):
        operation = fields.Str(
            required=True,
            validate=OneOf(_OPERATOR_FUNCTIONS, error="Not a permitted operation"),
        )
        value = AnyField(required=False, allow_none=True)

        @validates_schema
        def validate_operation_and_value(self, data, **kwargs):
            operation = data["operation"]
            value = data.get("value")
            if len(signature(operation).parameters) == 1:
                if value is not None:
                    raise ValidationError(
                        field_name="value",
                        message=f'No value should be provided for operation "{operation}"',
                    )
```

derekpierre replied:
✔️ - nice! Addressed in https://github.com/nucypher/nucypher/pull/3658/commits/91e373af2ca000748ed766b7bc07a341622753c0

manumonti commented on nucypher/policy/conditions/lingo.py:
```diff
@@ -291,18 +296,180 @@ def __init__(self, operand: Condition):
+        result = variable_value
+        for operation in operations:
+            result = operation.calc(result)
+        return result
```
I'm wondering if we can find a more descriptive name for this function.

- calc_sequentially_form_list
- apply_sequential_operations
- calc_operation_list
- apply_operations_sequentially

WDYT?

derekpierre replied:
✔️  - Ended up using `evaluate_operations` - https://github.com/nucypher/nucypher/pull/3658/commits/5b927a7ffc404c43508dde8937bf7384dc799b81

manumonti commented on tests/unit/conditions/test_variable_operation.py:
```diff
@@ -0,0 +1,161 @@
+        4,
+        ["Proper", "preparation", "prevents", "poor", "performance"],
+        "performance",
+    ),  # -- Ray Lewis
```
🤓

manumonti commented on tests/unit/conditions/test_signing_condition.py:
```diff
@@ -848,3 +857,160 @@ def test_signing_object_abi_attribute_condition_lingo_json_serialization(
+                condition=json_condition,
+                operations=[
+                    VariableOperation(operation="sum"),
+                    VariableOperation(operation="*=", value=10**18),  # convert to wei
```
```suggestion
                    VariableOperation(operation="ethToWei"),
```

derekpierre replied:
✔️

manumonti reviewed (dismissed):
I had some suggestions 🤏

derekpierre committed:
Handle the case where floating points are included in lists, or dictionaries and need to be converted to decimals for operations. Ensure that decimal values are converted back to floats where necessary for the final result.

derekpierre committed:
Rename calc_from_list to evaluate_operations, and _calc to _evaluate.

derekpierre committed:
Utilize ethToWei operation for test.

derekpierre committed:
Better strategy for detecting unary operations.

Co-authored-by: Manuel Montenegro <manuel@nucypher.com>

derekpierre committed:
Merge pull request #9 from cygnusv/variable-operations

Variable operations pt2

derekpierre committed:
Appease linter.

derekpierre commented:
> We have to be very careful with what operations we allow here. There's many potential traps (e.g. floating point precision issues, overloaded operators, etc) that we need to weigh very carefully. Please see my branch https://github.com/cygnusv/nucypher/tree/variable-operations, and in particular the last 2 commits with some examples.

Good catch, and thanks for the thoughtful changes.

It's good that we now use `Decimal` for calculations, but we have to be careful because it is really an internal structure. We can't totally get rid of precision issues, but we can limit them. The concern with returning `Decimal` values from operations as a result is that:
1. Decimal is really for internal calculations
2. Decimal cannot be used with `ast.literal_eval()` which is used for value comparison
3. Decimal cannot be specified as a comparator value in ReturnValueTest
4. Decimal is not JSON serializable so someone using the floating point result in a JSON would not be able do so if left as Decimal and not float

So I made some updates on top of the PR. Take a look and let me know what you think.  There is some loss of precision when converting from Decimal back to float at the very end of all of the operations, but I don't see that being an issue for the most part given the level of precision that can be achieved. There is also the option for the user to perform the `round` operation on results to given precision if necessary.

derekpierre committed:
Code cleanup based on RFCs from review.

derekpierre committed:
Handle conversion of operation parameter value outside of constructor, since if used within constructor the modified value gets serialized which is not what we want.
Add test to condition lingo to make sure that this is properly tested - original error showed up as part of taco-web integration test.

KPrasch reviewed (approved):
👍🏻 looking good after the removal of exponentiation!

manumonti reviewed (approved):
Looking good! 🔝
